transports/onion/src/lib.rs

```diff
@@ -0,0 +1,233 @@
+#![doc(html_logo_url = "https://libp2p.io/img/logo_small.png")]
+#![doc(html_favicon_url = "https://libp2p.io/img/favicon.png")]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+//! Tor based transport for libp2p. Connect through the Tor network to TCP listeners.
+//!
+//! Main entrypoint of the crate: [OnionTransport]
+//!
+//! ## Example
+//! ```no_run
+//! # use async_std_crate as async_std;
+//! # use libp2p_core::Transport;
+//! # async fn test_func() -> Result<(), Box<dyn std::error::Error>> {
+//! let address = "/dns/www.torproject.org/tcp/1000".parse()?;
+//! let mut transport = libp2p_onion::AsyncStdNativeTlsOnionTransport::bootstrapped().await?;
+//! // we have achieved tor connection
+//! let _conn = transport.dial(address)?.await?;
+//! # Ok(())
+//! # }
+//! # async_std::task::block_on(async { test_func().await.unwrap() });
+//! ```
+
+use std::sync::Arc;
+use std::{marker::PhantomData, pin::Pin};
+
+use address::{dangerous_extract_tor_address, safe_extract_tor_address};
+use arti_client::{DataStream, TorClient, TorClientBuilder};
+use futures::{future::BoxFuture, FutureExt};
+use libp2p_core::{transport::TransportError, Multiaddr, Transport};
+use tor_rtcompat::Runtime;
+
+mod address;
+mod provider;
+
+#[cfg(feature = "tokio")]
+#[cfg_attr(docsrs, doc(cfg(feature = "tokio")))]
+#[doc(inline)]
+pub use provider::OnionTokioStream;
+
+use provider::OnionStream;
+
+pub type OnionError = arti_client::Error;
+
+#[derive(Clone)]
+pub struct OnionTransport<R: Runtime, S> {
+    // client is in an Arc, because wihtout it the Transport::Dial method can't be implemented,
+    // due to lifetime issues. With the, eventual, stabilization of static async traits this issue
+    // will be resolved.
+    client: Arc<TorClient<R>>,
+    /// The used conversion mode to resolve addresses. One probably shouldn't access this directly.
+    /// The usage of [OnionTransport::with_address_conversion] at construction is recommended.
+    pub conversion_mode: AddressConversion,
+    phantom: PhantomData<S>,
+}
+
+/// Configure the onion transport from here.
+pub type OnionBuilder<R> = TorClientBuilder<R>;
+
+/// Mode of address conversion. Refer tor [arti_client::TorAddr](https://docs.rs/arti-client/latest/arti_client/struct.TorAddr.html) for details.
+#[derive(Debug, Clone, Copy, Hash, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub enum AddressConversion {
+    /// uses only dns for address resolution (default)
```

Comment:
```suggestion
    /// Uses only dns for address resolution (default).
```

For the sake of consistency, I would prefer comments to (a) be full sentences and (b) start with a capital letter.